Add Blueoil CLI

odoku opened the pull request:
<!--- 👉👉👉 Provide a general summary of your changes. 👈👈👈 -->

## Motivation and Context
<!--- 💭💡💭 Why is this change required? What problem does it solve? 💭💡💭 -->
<!--- 🔗 If it fixes an open issue, please link to the issue here. 🔗 -->

It was hard to use that it became an independent command for each task.
I thought it would be more convenient to aggregate into one command and call it as a subcommand.

## Description
<!--- 📝🎯📝 Describe your changes in detail. 📝🎯📝 -->

I designed the command as follows.

```console
# blueoil --help
Usage: blueoil [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  convert  Convert trained model to binary files.
  init     Generate blueoil config.
  predict  Predict by using trained model.
  train    Run training.
```

```console
# blueoil init --help
Usage: blueoil init [OPTIONS]

  Generate blueoil config.

Options:
  -o, --output TEXT  Path of generated configuration file.
  --help             Show this message and exit.
```

```console
# blueoil train --help
Usage: blueoil train [OPTIONS]

  Run training.

Options:
  -c, --config TEXT         Path of config file.  [required]
  -e, --experiment_id TEXT  ID of this training.
  --help                    Show this message and exit.
```

```console
# blueoil convert --help
Usage: blueoil convert [OPTIONS]

  Convert trained model to binary files.

Options:
  -e, --experiment_id TEXT  ID of this experiment.  [required]
  -p, --checkpoint TEXT     Checkpoint name. e.g. save.ckpt-10001  [required]
  -t, --template TEXT       Path of output template directory.
  --help                    Show this message and exit.
```

```console
# blueoil predict --help
Usage: blueoil predict [OPTIONS]

  Predict by using trained model.

Options:
  -i, --input TEXT                Directory containing predicted images.
								  [required]
  -o, --output TEXT               Directory to output prediction result.
								  [required]
  -e, --experiment_id TEXT        ID of this experiment.  [required]
  -p, --checkpoint TEXT           Checkpoint name. e.g. save.ckpt-10001
								  [required]
  --save_images / --no_save_images
								  Flag of saving images. Default is True.
  --help                          Show this message and exit.
```

As usual you can also use `blueoil.sh` method.

## How has this been tested?
<!--- 🙆👌🙆 Please describe in detail how you tested your changes. 🙆👌🙆 -->
<!--- Include details of your testing environment, details of your manual tests, snippet code or commands of your manual tests, table of experiments results metrics, tests ran to see how your change affects other areas of the code, etc. -->
<!--- If you have experiments report documents please link to here. -->

I confirmed that the following operations are completed.

```console
git clone --recursive -b blueoil-cli git@github.com:odoku/blueoil.git

cd blueoil
./docker_build.sh

mkdir dataset
cd dataset
wget https://s3-ap-northeast-1.amazonaws.com/leapmind-public-storage/datasets/cifar.tgz
tar xvf cifar.tgz

cd ../
docker run \
	--rm \
	-it \
	--runtime=nvidia \
	-e PYTHONPATH=/home/blueoil:/home/blueoil/lmnet:/home/blueoil/dlk/python/dlk \
	-e OUTPUT_DIR=/home/blueoil/saved \
	-v $(pwd)/config:/home/blueoil/config \
	-v $(pwd)/dataset:/home/blueoil/dataset \
	-v $(pwd)/saved:/home/blueoil/saved \
	-v $(pwd)/result:/home/blueoil/result \
	$(id -un)_blueoil:local_build \
	/bin/bash

./blueoil/cli.py init  -o /home/blueoil/config/config.yml
./blueoil/cli.py train -c /home/blueoil/config/config.yml
./blueoil/cli.py convert -e train_20190115130740 -p save.ckpt-781
./blueoil/cli.py predict -i /home/blueoil/dataset/cifar/test/dog -o /home/blueoil/result -e train_20190115130740 -p save.ckpt-781
```

## Types of changes
<!--- 🏁🎌🏁 What types of changes does your code introduce? Put an `x` in all the boxes that apply: 🏁🎌🏁 -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- ✅✅✅ Go over all the following points, and put an `x` in all the boxes that apply. ✅✅✅ -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

<!--- TODO(wakisaka): After decide our code style, add this.
- [ ] My code follows the code style of this project.
-->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

lm-konda commented:
@odoku Thank you for your contribution.

To LM members,
how do we manage jenkins test?
I think these kinds of PR don't pass jenkins test from it's structure (currently seems to stop).

lm-konda commented:
Although this mechanism itself is good, I think that the ease of operation which is the thought of blueoil is impaired. For example, we have used docker in implementations up to now, but it should have been an implementation that almost never considers docker. I think that the mechanism of this time is not so good from that point of view. How about that?

odoku commented:
Sorry, I made some changes to PR.
Changed to specify checkpoint name only because the restore_path option was inconvenient.
I changed the description of RP and tested again.

odoku commented on setup.cfg:
```diff
@@ -1,6 +1,6 @@
 name = blueoil
-version = 0.1
+version = 0.2.1
 url = https://leapmind.io/
```
@kchygoe 
I think, setup.py does not appear to be maintained.
It seems that it does not seem to be a problem, so I will change it at once.

odoku commented:
@lm-konda 

You can still use blueoil.sh as usual.

This correspondence is to make Blueoil's correct Docker image.
I am writing a Dockerfile now, but when completed it will be available as follows.
Although this looks inconvenient at first glance, it is very convenient when using it from a program (For example DeLTA-Lite).
And if you can install lmnet and dlk correctly, you can register with PYPI and install it with pip as well.

I don't think convenience is being impaired. Because no one function has been impaired.

```console
docker run --rm --runtime=nvidia \
	-it \
	-v $(pwd)/config:/blueoil/config \
	-v $(pwd)/datasets:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1 init -o /blueoil/config/config.yml

docker run --rm --runtime=nvidia \
	-v $(pwd)/config:/blueoil/config \
	-v $(pwd)/datasets:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1 train -c /blueoil/config/config.yml

docker run --rm --runtime=nvidia \
	-v $(pwd)/datasets:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	blueoil:0.2.1 convert -e train_20190115070835 -p save.ckpt-2343

docker run --rm --runtime=nvidia \
	-v $(pwd)/datasets:/blueoil/dataset \
	-v $(pwd)/saved:/blueoil/saved \
	-v $(pwd)/result:/blueoil/result \
	blueoil:0.2.1 predict -i /blueoil/dataset/cifar/test/dog -o /blueoil/result -e train_20190115070835 -p save.ckpt-2343
```

ruimashita commented:
@nlpng @lm-jira 
Please review this PR

djKooks commented on blueoil/blueoil_init.py:
```diff
@@ -187,15 +190,18 @@ def image_size_filter(raw):
-    config_file = blueoil_config['model_name'] + ".yml"
-    with open(config_file, 'w') as fp:
+
+    if not output:
```
How about checking format of `output` here, instead of just checking this is `None` or not? Such as:
```python
# this is sudo code...may cannot work
if not output or output.endswith('.yml') == False:
```

odoku replied:
@djKooks 
I looked at the loading process of the config file and I did not specify extensions, so I thought that it was not necessary at init time.

https://github.com/blue-oil/blueoil/blob/4e6abd97f5f740af89484c08a7d69934e58c2d56/blueoil/generate_lmnet_config.py#L77-L96

ruimashita commented:
Jenkins, ok to test

ruimashita commented on blueoil/blueoil_convert.py:
```diff
@@ -46,10 +46,11 @@ def get_export_directory(experiment_id, restore_path):
 
-    template_dir = "/home/blueoil/output_template"
+    # TODO: Remove the hard-coded path.
+    template_dir = "/home/blueoil/output_template" if not output_template_dir else output_template_dir
```
Can you use TEMPLATE_DIR instead of  `/home/blueoil/output_template` ?

odoku replied:
@ruimashita

What `TEMPLATE_DIR` you are talking about is it?
https://github.com/blue-oil/blueoil/blob/741a5bd93e2210c64167dddc22078ee2c50d72dc/blueoil/vars.py#L19-L20

This is the template path of the blueoil/lmnet config file.
It is the template of the conversion result (run.py, usb_camera_demo.py, etc...) that we must specify with `output_template_dir`.

ruimashita replied:
@odoku 
Oh, I mistaken 😭 `TEMPLATE_DIR` is not `/home/blueoil/output_template` .
In order to remove hard-code, Could you create new constant like `OUTPUT_TEMPLATE_DIR` ?

odoku replied:
@ruimashita I fixed it. Please check again! https://github.com/blue-oil/blueoil/pull/105/commits/cdf77ce36eff53a31b6961751a183a8a748b7502

ruimashita replied:
@odoku Thanks

ruimashita commented:
ok to test

ruimashita commented:
@odoku Thanks 😄
